overlord/hookstate/hookmgr.go

```diff
@@ -296,16 +327,45 @@ func runHookAndWait(snapName string, revision snap.Revision, hookName, hookConte
 		close(hookCompleted)
 	}()
 
-	select {
-	// Hook completed; it may or may not have been successful.
-	case <-hookCompleted:
-		return buffer.Bytes(), hookError
+	var abortOrTimeoutError error
+	dyingCh := tomb.Dying()
+out:
+	for {
```

Comment:
The logic would probably be more understandable without this loop. Considering a timeout or abort only makes sense once, and if that happens we should get into an error mode. If that doesn't happen we should get into a success mode. All of these are a single tree of decisions, instead of a loop. As it is I'm finding it hard to keep in mind all possible outcomes of having different events in different orders and trying to prove them right.